File: kiro26/geoip_last.py
"""
geoip_last.py — reads /var/log/wtmp, prints GeoIP city/country per login IP.

Usage (CLI):  python geoip_last.py [wtmp_file]
Usage (lib):  from geoip_last import iter_logins, geoip_lookup
"""

import ipaddress
import socket
import struct
import sys
from functools import lru_cache
from urllib.request import urlopen
from urllib.error import URLError
import json

# ── utmp struct (Linux/macOS) ─────────────────────────────────────────────────
# matches struct utmp in <utmp.h>
_FMT  = "hi32s32s32s256sHHiI4i20s"
_SIZE = struct.calcsize(_FMT)
USER_PROCESS = 7


def iter_logins(path: str = "/var/log/wtmp"):
    """Yield IP strings for every USER_PROCESS entry that has a remote host."""
    with open(path, "rb") as f:
        while chunk := f.read(_SIZE):
            if len(chunk) < _SIZE:
                break
            fields = struct.unpack(_FMT, chunk)
            ut_type = fields[0]
            ut_host = fields[5].rstrip(b"\x00").decode(errors="ignore")
            if ut_type != USER_PROCESS or not ut_host:
                continue
            ip = _resolve(ut_host)
            if ip and not _is_private(ip):
                yield ip


def _resolve(host: str) -> str | None:
    try:
        ipaddress.ip_address(host)
        return host
    except ValueError:
        pass
    try:
        return socket.gethostbyname(host)
    except socket.gaierror:
        return None


def _is_private(ip: str) -> bool:
    try:
        return ipaddress.ip_address(ip).is_private
    except ValueError:
        return True


@lru_cache(maxsize=512)
def geoip_lookup(ip: str) -> tuple[str, str]:
    """Return (city, country) for an IP via ip-api.com (no key required)."""
    try:
        url = f"http://ip-api.com/json/{ip}?fields=status,city,country"
        with urlopen(url, timeout=5) as r:
            data = json.loads(r.read())
        if data.get("status") == "success":
            return data.get("city", "?"), data.get("country", "?")
    except (URLError, json.JSONDecodeError):
        pass
    return "?", "?"


# ── CLI ───────────────────────────────────────────────────────────────────────
def main():
    path = sys.argv[1] if len(sys.argv) > 1 else "/var/log/wtmp"
    try:
        ips = list(iter_logins(path))
    except FileNotFoundError:
        print(f"Error: {path} not found", file=sys.stderr)
        sys.exit(1)

    if not ips:
        print("No remote logins found.")
        return

    print(f"{'IP':<45} {'City':<22} Country")
    print("-" * 75)
    for ip in ips:
        city, country = geoip_lookup(ip)
        print(f"{ip:<45} {city:<22} {country}")


if __name__ == "__main__":
    main()

File: kiro26/geoip_last.cpp
// geoip_last.cpp — reads /var/log/wtmp, prints GeoIP city/country per login IP
// Build: g++ -std=c++26 -o geoip_last geoip_last.cpp -lcurl
// Usage: ./geoip_last [wtmp_file]

#include <curl/curl.h>
#include <utmp.h>

#include <arpa/inet.h>
#include <netdb.h>

#include <cstring>
#include <format>
#include <fstream>
#include <iostream>
#include <print>
#include <string>
#include <unordered_map>

// ── libcurl write callback ────────────────────────────────────────────────────
static size_t write_cb(char* ptr, size_t, size_t nmemb, std::string* out) {
    out->append(ptr, nmemb);
    return nmemb;
}

// ── Minimal JSON field extractor (no dependency) ──────────────────────────────
static std::string json_field(const std::string& json, const std::string& key) {
    auto pos = json.find(std::format("\"{}\"", key));
    if (pos == std::string::npos) return "?";
    pos = json.find(':', pos);
    if (pos == std::string::npos) return "?";
    ++pos;
    while (pos < json.size() && (json[pos] == ' ' || json[pos] == '"')) ++pos;
    auto end = json.find_first_of("\",}", pos);
    return end == std::string::npos ? "?" : json.substr(pos, end - pos);
}

// ── GeoIP lookup via ip-api.com ───────────────────────────────────────────────
struct GeoInfo { std::string city, country; };

static GeoInfo geoip(const std::string& ip) {
    CURL* curl = curl_easy_init();
    if (!curl) return {"?", "?"};

    std::string body;
    auto url = std::format("http://ip-api.com/json/{}?fields=status,city,country", ip);
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &body);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
    curl_easy_perform(curl);
    curl_easy_cleanup(curl);

    if (json_field(body, "status") != "success") return {"?", "?"};
    return {json_field(body, "city"), json_field(body, "country")};
}

// ── Resolve hostname → IP string ──────────────────────────────────────────────
static std::string resolve(const char* host) {
    // Already an IP?
    struct in_addr a{}; struct in6_addr a6{};
    if (inet_pton(AF_INET, host, &a) == 1 || inet_pton(AF_INET6, host, &a6) == 1)
        return host;

    addrinfo hints{}, *res{};
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    if (getaddrinfo(host, nullptr, &hints, &res) != 0) return {};

    char buf[INET6_ADDRSTRLEN]{};
    if (res->ai_family == AF_INET)
        inet_ntop(AF_INET, &reinterpret_cast<sockaddr_in*>(res->ai_addr)->sin_addr, buf, sizeof buf);
    else
        inet_ntop(AF_INET6, &reinterpret_cast<sockaddr_in6*>(res->ai_addr)->sin6_addr, buf, sizeof buf);
    freeaddrinfo(res);
    return buf;
}

// ── main ──────────────────────────────────────────────────────────────────────
int main(int argc, char* argv[]) {
    const char* path = argc > 1 ? argv[1] : "/var/log/wtmp";

    std::ifstream f(path, std::ios::binary);
    if (!f) { std::println(stderr, "Cannot open {}", path); return 1; }

    curl_global_init(CURL_GLOBAL_DEFAULT);

    std::unordered_map<std::string, GeoInfo> cache;
    std::println("{:<45} {:<22} {}", "IP", "City", "Country");
    std::println("{}", std::string(75, '-'));

    utmp entry{};
    while (f.read(reinterpret_cast<char*>(&entry), sizeof entry)) {
        // Only USER_PROCESS entries with a remote host (like last.c)
        if (entry.ut_type != USER_PROCESS) continue;
        const char* host = entry.ut_host;
        if (host[0] == '\0') continue;

        std::string ip = resolve(host);
        if (ip.empty()) continue;

        // Skip private/loopback addresses
        if (ip.starts_with("127.") || ip.starts_with("10.") ||
            ip.starts_with("192.168.") || ip == "::1") continue;

        if (!cache.contains(ip)) {
            std::print("Looking up {}...\r", ip);
            std::cout.flush();
            cache[ip] = geoip(ip);
        }
        const auto& [city, country] = cache[ip];
        std::println("{:<45} {:<22} {}", ip, city, country);
    }

    curl_global_cleanup();
}
